piptools/resolver.py

```diff
@@ -87,14 +88,17 @@ def resolve(self, max_rounds=10):
 
     def _check_constraints(self):
         for constraint in chain(self.our_constraints, self.their_constraints):
-            if constraint.link is not None and not constraint.editable:
-                msg = ('pip-compile does not support URLs as packages, unless they are editable '
-                       '(perhaps add -e option?)')
-                raise UnsupportedConstraint(msg)
+            if constraint.link and constraint.link.scheme not in vcs.all_schemes:
```

Comment:
The proper way of adding this would be to contribute this convenience method on `Link` to pip itself, have that PR pulled, wait for the pip release, and then bump the requirement to the latest pip. (See also the comment below about not relying too much on pip's implementation details.)